src/pykeen/nn/emb.py

```diff
@@ -0,0 +1,173 @@
+# -*- coding: utf-8 -*-
+
+"""Embedding modules."""
+
+import functools
+from typing import Any, Callable, Mapping, Optional, TypeVar
+
+import torch
+import torch.nn
+from torch import nn
+
+__all__ = [
+    'RepresentationModule',
+    'Embedding',
+    'Initializer',
+    'Normalizer',
+    'Constrainer',
+]
+
+
+class RepresentationModule(nn.Module):
+    """A base class for obtaining representations for entities/relations."""
+
+    def forward(
+        self,
+        indices: Optional[torch.LongTensor] = None,
+    ) -> torch.FloatTensor:
+        """Get representations for indices.
+
+        :param indices: shape: (m,)
+            The indices, or None. If None, return all representations.
+
+        :return: shape: (m, d)
+            The representations.
+        """
+        raise NotImplementedError
+
+    def reset_parameters(self) -> None:
+        """Reset the module's parameters."""
+
+    def post_parameter_update(self):
+        """Apply constraints which should not be included in gradients."""
+
+
+# FIXME: It has to be more than one type...
+TensorType = TypeVar("TensorType", torch.Tensor, torch.FloatTensor)
+Initializer = Callable[[TensorType], None]
+Normalizer = Callable[[TensorType], TensorType]
+Constrainer = Callable[[TensorType], TensorType]
+
+
+class Embedding(RepresentationModule):
+    """Trainable embeddings.
+
+    This class provides the same interface as :class:`torch.nn.Embedding` and
+    can be used throughout PyKEEN as a more fully featured drop-in replacement.
+    """
+
+    def __init__(
+        self,
+        num_embeddings: int,
+        embedding_dim: int,
+        initializer: Optional[Initializer] = None,
+        initializer_kwargs: Optional[Mapping[str, Any]] = None,
+        normalizer: Optional[Normalizer] = None,
+        normalizer_kwargs: Optional[Mapping[str, Any]] = None,
+        constrainer: Optional[Constrainer] = None,
+        constrainer_kwargs: Optional[Mapping[str, Any]] = None,
+    ):
+        super().__init__()
+
+        if initializer is None:
+            initializer = nn.init.normal_
```

Comment:
does it make sense to have a default for this?

Reply:
This is the default of torch.nn.Embedding, cf. https://pytorch.org/docs/stable/generated/torch.nn.Embedding.html#torch.nn.Embedding

Reply:
Okay!